### DISCUSSION TITLE
query string all or none? (query string nested struct)

### DISCUSSION THREAD

**Discussion by zzzz465** [actix/actix-web#2877] [Q&A] [answered]
Hi, I'm trying to make a GET handler that use query strings.
the API can be called two ways.
1. call without any query strings.
2. call with fully provided query strings. partial data will not be allowed.

```rs
#[derive(Serialize, Deserialize, Debug)]
pub struct GetQuery {
    train_id: Option<String>,
    departure: Option<String>,
}
```

with this struct, in a GET handler I have to check whether GetQuery is partially filled or not (`train_id`, `departure` both should be none or some)
since I wanted to delegate this task to actix_web. I could write struct like this second form.

```rs
#[derive(Serialize, Deserialize, Debug)]
pub struct GetQuery {
    data: Option<Data>
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Data {
    train_id: String,
    departure: String
}
```
then this raises a question. query string only allows primitive key-value data. how actix_web will treat this?

**Comment by zzzz465** [comment]
although it doesn't directly answer the question, I found that nested struct can be flattened using `#[serde(flatten)]` which makes possible to put everything on a 1-depth key-value. so I can fit data into query string.

**Comment by robjtede** [comment, marked as answer]
For an all-or-nothing approach, you can use an Option extractor.

```rust

#[derive(Serialize, Deserialize, Debug)]
pub struct Data {
    train_id: String,
    departure: String
}

async fn handler(query: Option<web::Query<Data>>) { ... }
```

**Reply by zzzz465** [reply]
great! thank you.
